### ISSUE TITLE
App is blocked by Google

### ISSUE THREAD

**Issue by OZoneGuy** [insanum/gcalcli#580]
I tried using `gcalcli list` for the first time and it opened a browser tab asking for verification. After choosing the account I got the following page:

![image](https://user-images.githubusercontent.com/5082774/104369188-0dc51680-54eb-11eb-8680-2f7fd6a586a3.png)

Not sure if this is intended or known. I tried this with my personal account and my University's account and I got the same result in both scenarios. I tried using the `--noauth_local_webserver` option and got the same result after logging in.

Not sure how to solve it, and I couldn't find any similar issues online.

I am tried using both `gcalcli` and `gcalcli-git` found on the AUR.

**Comment by katacarbix** [comment]
Your best bet is to follow the steps here to get your own API keys: https://github.com/insanum/gcalcli#login-information

Though I have a problem with that where it keeps logging me out and making me re-authenticate every week or so. I may need to get it verified for it to stick.

**Comment by OZoneGuy** [comment]
Maybe. I tried creating an API key like the README suggests, but then I get this error page:

![image](https://user-images.githubusercontent.com/5082774/104464252-b5405880-5580-11eb-8afc-651dcf145869.png)

I haven't used the Google auth API in a long time, so I don't know what has changed or what needs to be done.

**Comment by katacarbix** [comment]
In the Cloud Console go to OAuth Consent Screen -> Edit App -> Click continue twice to get to "Test users" and you should be able to add your email.

**Comment by aurisnoctis** [comment]
tldr; OAuth worked for me, but only in Chromium, not in Firefox.

Steps, loosely following comment https://github.com/insanum/gcalcli/issues/497#issuecomment-592948891

#### 1. **Google Developers Console** https://console.developers.google.com
* create a new project, e.g., `gcalcli`

#### 2. **+ Enable APIs and Services"**
*  enable `Google Calendar API"
* go back to project `gcalcli` in Developers Console

#### 3. **Credentials** section in left menu
* "+ Create Credentials" > `OAuth client ID`
* Application type `Desktop App`
* name `gcalcli`
* **The credentials produced here are needed in step 5**

#### 4. **OAuth consent screen** section in left menu
* I was restricted to choose "External users"
* OAuth consent screen
    * App name": e.g., gcalcli
    * User support email: e.g., put your own gmail address
    * App domain: leave all blank
    * Developer contact information: put your own gmail address
* Scopes: choose all calendar API related stuff plus more if necessary - I did trial and error here ...
*  Test users: again I put my own gmail address here
* Summary: click "Back to dashboard" at bottom
* leave `Publishing status` as `Testing` so your own gcalcli is not published to the world

#### 5. call gcalcli from command line
* Use credentials from step 3 here in cli command, e.g. for Google agenda view: `gcalcli --client_id "your client ID ending in ...apps.googleusercontent.com" --client_secret "your client secret" agenda`
* as my default browser is Firefox I had to copy the URL after `Your browser has been opened to visit:` (shown in shell after above command) and open it in chromium > follow steps and give permissions to `gcalcli`
* in shell, Google calendar agenda is shown

SO easy! :'D

**Comment by aReaSa** [comment]
I tried what @aurisnoctis said. However, I couldn't be successful. Is there any other way around to make this work out?
Thank you.

![image](https://user-images.githubusercontent.com/26048237/108396008-a94a5500-7216-11eb-95de-39a67e79d1dc.png)

**Comment by katacarbix** [comment]
I'm using Firefox and it gives me the option to continue with a warning:
![gcalcli + Firefox](https://user-images.githubusercontent.com/1340784/108413522-bf451f00-71f0-11eb-8dd6-fd64795ce838.png)

**Comment by jcmayoral** [comment]
Thanks @aurisnoctis Really helpful.

**Comment by avanavana** [comment]
I had been using this awesome app for about a month, and suddenly today, Google tells me that the app is blocked, and there's nothing I can do to get it authenticated.  I had already set it up with my own GCP OAuth credentials, using .gcalclirc and it was working fine.   Also, "Allow less secure apps" setting on my Google Account is "ON".   I don't understand, I've been developing Google Cloud Platform apps for a few years and I was always given the option to "continue" and "allow" the app, now there is just no option, only a "Blocked" message.  Any ideas?

**Comment by avanavana** [comment]
I solved this issue using the following resources:

1. Make sure to turn on your Google account's setting for "[Allow less secure apps](https://www.google.com/settings/security/lesssecureapps)"
2. Visit Google's "[Display Unlock Captcha](https://accounts.google.com/DisplayUnlockCaptcha)" page and click "Continue" to remove the security block.

**Comment by DavidTre07** [comment]
Hello all,
what about generating a token from this page ? https://myaccount.google.com/apppasswords
Could it solve this problem ? I think gcalcli is not ready to use such token but it could be a great improvment.

**Comment by andersfylling** [comment]
Same issue. I'm trying to read my schedule from my work account, where I can't change any of the above settings.

**Comment by aristosv** [comment]
I tried this guide for creating the client id and client secret and it worked flawlessly. I didn't even have to run "DisplayUnlockCaptcha" nor enable Less Secure Apps.

https://github.com/aristosv/google_auth

**Comment by Mellbourn** [comment]
I think the readme of this project should state a bit more clearly that you may have to do a 5+ step guide to authorize this app. Currently, it implies that only OAuth2 will be needed and anything else is optional.

**Comment by eastpole** [comment]
Just chiming in here that yes, the problem is still blocking simple usage of gcalcli, but on the other hand, the complicated process above does still work when applied with some flexibility. Google keeps changing the dashboards of the relevant API management, so it's hard to feel confident that you are doing the right thing. :)   Thank you, @AurisNoctis

**Comment by sloanlance** [comment]
Additional information…

I've been using gcalcli for about a week.  I had authorized it to access one of my two Google accounts (my personal account) following the steps in the documentation.  It was an awkward process, but I somewhat understand the complexities of working with Google's security.

After using gcalcli for about a week, today I tried to use it to access the calendar belonging to my other Google account (my account in my employer's enterprise).  In Google Calendar, long ago I gave each of my Google accounts mutual access to their calendars.  I tried to use the gcalcli authorization for my personal account to add an event to my work account.  At first, I got the following stack trace…

```txt
Traceback (most recent call last):
  File "/Users/lsloan/homebrew/bin/gcalcli", line 33, in <module>
    sys.exit(load_entry_point('gcalcli==4.3.0', 'console_scripts', 'gcalcli')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/gcalcli/cli.py", line 181, in main
    gcal.QuickAddEvent(
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/gcalcli/gcal.py", line 1294, in QuickAddEvent
    new_event = self._retry_with_backoff(
                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/gcalcli/gcal.py", line 112, in _retry_with_backoff
    return method.execute()
           ^^^^^^^^^^^^^^^^
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
    return wrapped(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/googleapiclient/http.py", line 923, in execute
    resp, content = _retry_request(
                    ^^^^^^^^^^^^^^^
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/googleapiclient/http.py", line 191, in _retry_request
    resp, content = http.request(uri, method, *args, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/oauth2client/transport.py", line 186, in new_request
    credentials._refresh(orig_request_method)
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/oauth2client/client.py", line 761, in _refresh
    self._do_refresh_request(http)
  File "/Users/lsloan/homebrew/Cellar/gcalcli/4.3.0_3/libexec/lib/python3.11/site-packages/oauth2client/client.py", line 819, in _do_refresh_request
    raise HttpAccessTokenRefreshError(error_msg, status=resp.status)
oauth2client.client.HttpAccessTokenRefreshError: invalid_grant: Token has been expired or revoked.
```

Thinking that it might be a temporary problem, I tried again.  This time, gcalcli opened the following address in a web browser…

```txt
https://accounts.google.com/o/oauth2/v2/auth?client_id=«REDACTED».apps.googleusercontent.com&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2F&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar&access_type=offline&response_type=code
```

In the Google page that opened, I selected my account name when prompted.  The next page showed me the familiar message others have mentioned…

```txt
This app is blocked
This app tried to access sensitive info in your Google Account. To keep your account safe, Google blocked this access.
```

I'm not certain whether this was caused by my token being expired/revoked as mentioned in the first error message or maybe it was triggered by trying to access my work account calendar from my personal account.

**Comment by jamespo** [comment]
I've made some modifications in a fork using newer auth methods. Download a creds.json & put it in ~/.config/gcalcli/ - details in the update README.

I'm not providing support, but you can see if it works for you (it works for me).

Be warned this is a quick 30 minute hack, I suggest creating a virtualenv and installing it in there. Also it used my existing ~/.gcalcli_oauth file (not sure why currently) so I moved it to start a new auth flow.

I suggest inspecting the diffs, no guarantees this won't wipe all your appointments!

https://github.com/jamespo/gcalcli

**Comment by dbarnett** [comment]
Alright, I'm removing the broken client ID from the app entirely in #707 while making the setup process a little friendlier. I'm pretty sure that takes care of all loose ends here, but please reach out with a fresh issue if you're still seeing problems not related to that broken client ID.
